service/test/BatchServerTest.cpp

```diff
@@ -359,6 +461,263 @@ TEST_F(BatchServerTest, run_batch_write_empty)
+        .WillOnce(Return(127))  // a "random" number
+        .RetiresOnSaturation();
+
+    GEOPM_EXPECT_THROW_MESSAGE(
+        m_batch_server->run_batch(),
+        GEOPM_ERROR_RUNTIME,
+        "BatchServerImp::run_batch(): Received unknown response from client: "
+    );
+}
+
+/**
+ * @throws geopm Exception to simulate failure of read(2) system call in BatchStatusImp::receive_message()
+ */
+ACTION(batch_status_read_EINTR)
+{
+    throw geopm::Exception(
+        "BatchStatusImp: System call failed: " "read(2)",
+        EINTR,
+        __FILE__,
+        __LINE__
+    );
+}
+
+/**
+ * @test Testing BatchStatus::receive_message() to throw an exception inside of BatchServerImp::read_message()
+ *       This is needed to activate the catch blocks in BatchServerImp::read_message() and BatchServerImp::run_batch()
+ */
+TEST_F(BatchServerTest, receive_message_exception)
+{
+    InSequence sequence;
+
+    int idx = 0;
+    std::vector<double> result = {240.042, 250.052};
+
+    for (const auto &request : m_signal_config) {
+        EXPECT_CALL(*m_pio_ptr, push_signal(request.name, request.domain,
+                                            request.domain_idx))
+            .WillOnce(Return(idx))
+            .RetiresOnSaturation();
+        ++idx;
+    }
+
+    for (const auto &request : m_control_config) {
+        EXPECT_CALL(*m_pio_ptr, push_control(request.name, request.domain,
+                                             request.domain_idx))
+            .WillOnce(Return(idx))
+            .RetiresOnSaturation();
+        ++idx;
+    }
+
+    // This is where it throws the message.
+    EXPECT_CALL(
+        *m_batch_status.get(),
+        receive_message()
+    )
+    .Times(1)
+    .WillRepeatedly(batch_status_read_EINTR());
+
+    GEOPM_EXPECT_THROW_MESSAGE(
+        m_batch_server->run_batch(),
+        EINTR,
+        "BatchStatusImp: System call failed: " "read(2)"
+    );
+}
+
+/**
+ * @throws geopm Exception to simulate failure of write(2) system call in BatchStatusImp::send_message()
+ */
+ACTION(batch_status_write_EINTR)
+{
+    throw geopm::Exception(
+        "BatchStatusImp: System call failed: " "write(2)",
+        EINTR,
+        __FILE__,
+        __LINE__
+    );
+}
+
+/**
+ * @test Testing BatchStatusImp::send_message() to throw an exception inside of BatchServerImp::write_message()
+ *       This is needed to activate the catch blocks in BatchServerImp::write_message() and BatchServerImp::run_batch()
+ */
+TEST_F(BatchServerTest, write_message_exception)
+{
+    InSequence sequence;
+
+    int idx = 0;
+    std::vector<double> result = {240.042, 250.052};
+
+    for (const auto &request : m_signal_config) {
+        EXPECT_CALL(*m_pio_ptr, push_signal(request.name, request.domain,
+                                            request.domain_idx))
+            .WillOnce(Return(idx))
+            .RetiresOnSaturation();
+        ++idx;
+    }
+
+    for (const auto &request : m_control_config) {
```

Comment:
```suggestion
    idx = 0;
    for (const auto &request : m_control_config) {
```

When pushing signals and controls, make sure the index is reset in between.  It doesn't matter for this test because you are not verifying the indices, but it will lead to confusing behavior if this code is used in other tests that do utilizes the indices.